omod/src/main/java/org/openmrs/module/pihcore/PihBasicAuthenticationScheme.java

```diff
@@ -0,0 +1,89 @@
+/**
+ * The contents of this file are subject to the OpenMRS Public License
+ * Version 1.0 (the "License"); you may not use this file except in
+ * compliance with the License. You may obtain a copy of the License at
+ * http://license.openmrs.org
+ * <p>
+ * Software distributed under the License is distributed on an "AS IS"
+ * basis, WITHOUT WARRANTY OF ANY KIND, either express or implied. See the
+ * License for the specific language governing rights and limitations
+ * under the License.
+ * <p>
+ * Copyright (C) OpenMRS, LLC.  All Rights Reserved.
+ */
+package org.openmrs.module.pihcore;
+
+import org.apache.commons.lang.StringUtils;
+import org.openmrs.Location;
+import org.openmrs.api.context.Context;
+import org.openmrs.api.context.ContextAuthenticationException;
+import org.openmrs.module.authentication.AuthenticationUtil;
+import org.openmrs.module.authentication.web.AuthenticationSession;
+import org.openmrs.module.authentication.web.BasicWebAuthenticationScheme;
+import org.openmrs.module.pihcore.metadata.Metadata;
+
+import javax.servlet.http.HttpServletRequest;
+import java.util.List;
+import java.util.Properties;
+
+import static org.openmrs.module.appui.UiSessionContext.LOCATION_SESSION_ATTRIBUTE;
+import static org.openmrs.module.emr.EmrConstants.COOKIE_NAME_LAST_SESSION_LOCATION;
+
+/**
+ * This expands on the BasicWebAuthenticationScheme to also handle collecting, validating, and setting sessionLocation
+ */
+public class PihBasicAuthenticationScheme extends BasicWebAuthenticationScheme {
+
+	public static final String SESSION_LOCATION = "sessionLocation";
+	public static final String LOCATION_REQUIRED = "locationRequired";
+
+	private boolean locationRequired = true;
+
+	@Override
+	public void configure(String schemeId, Properties config) {
+		super.configure(schemeId, config);
+		locationRequired = AuthenticationUtil.getBoolean(config.getProperty(LOCATION_REQUIRED), true);
+	}
+
+	@Override
+	public void beforeAuthentication(AuthenticationSession session) {
+		super.beforeAuthentication(session);
+		Location loginLocation = getLoginLocation(session.getHttpRequest());
+		if (loginLocation == null && locationRequired) {
+			throw new ContextAuthenticationException("mirebalais.login.error.locationRequired");
+		}
+	}
+
+	@Override
+	public void afterAuthenticationSuccess(AuthenticationSession session) {
+		super.afterAuthenticationSuccess(session);
+		Location loginLocation = getLoginLocation(session.getHttpRequest());
+		if (loginLocation != null) {
+			Context.getUserContext().setLocation(loginLocation);
+			session.setHttpSessionAttribute(LOCATION_SESSION_ATTRIBUTE, loginLocation.getLocationId());
+			session.setCookieValue(COOKIE_NAME_LAST_SESSION_LOCATION, loginLocation.getLocationId().toString());
+		}
+	}
+
```

Comment:
Because the OpenMRS AuthenticationScheme construct is not web-aware, and the authenticate method has no way to access the HTTP Session, I added some additional hooks into a WebAuthenticationScheme interface that allows for additional web-specific behavior to be included in an AuthenticationScheme that is executed before authentication, and after an authentication success or failure.  I'm using this to include the logic formerly in the LoginController post submission handler to retrieve the login location from the request, validate it, and set it in the various places (session, context, cookie).  NOTE:  This cookie is a _session_ cookie - meaning once the browser is closed it will be cleared.  If we actually intended this to be a persistent cookie with a longer TTL, we should set that, but AFAIK this is what we have had in place up until now

Reply:
Happy to discuss further @mogoodrich , but how would you aim to achieve this?  As I said, we have no easy way to do this in the authenticate(Credentials) method, which is what OpenMRS core supports.  We could add more information to the Credentials (eg. the location), but this doesn't really help because this still does not have access to the HttpSession for setting Location as a Session Attribute, or the HttpServletResponse for setting this Location as a Cookie, so we would still need these to be done outside of the authenticate method.  We could pass the entire AuthenticationSession (or just the Session and Response) in as part of the Credentials, but I am avoiding that because I think we need to keep the Credentials Serializable and light weight, and there are cyclical issues (Credentials are contained in the Session, so we can't really add the Session to the Credentials).

The solution above doesn't really "tack things on".  The alternative to the above (in design, but not in substance) would be to have a new authenticate method in a WebAuthenticationScheme base class which consumers would have to know to use in place of the existing one:

```
public Authenticated authenticate(Credentials credentials, AuthenticationSession session) {
   //  stuff to do before authentiation
   try {
       Authenticated authenticated = authenticate(credentials);
       // stuff to do after authentication succeeds
   }
   catch (Exception e) {
      // stuff to do after authentication fails
      if (!(e instanceof ContextAuthenticationException)) {
         e = new ContextAuthenticationException(e);
      }
      throw e; 
   }
}
```

If you have alternative or better ideas, let's please discuss.

Reply:
Oops, sorry I wasn't clear in my comment. I was saying that your new approach was better than the *previous* tacked on approach... not suggesting *your* approach was tacked on...  sorry again, I should have just said "nice work" :)